setup.cfg can also be a place to store the version

ewjoachim opened the pull request:
This is another implementation of #230. I'm not really satisfied because this is more or less copy/pasted code. (If we decide to go this way, I'll integrate the docs in this PR too).

Let me know what you prefer :)

mauritsvanrees commented:
Both approaches look okay at first glance. This one seems simpler, with less changes. I have not tried it out. Apparently `setup.cfg` could also contain `version = attr: src.VERSION`. I don't know if that maybe trips up the logic here.
I don't really have time now for a good look, and will be away for a week. And I think @reinout is on vacation but could be back soon.

ewjoachim commented:
After re-reading myself, this version needs the change in the regex pattern too. This regex pattern includes a digit so we're safe for version = attrs:

I'm totally ok if we leave this PR sleep for a few weeks :) Enjoy your time off !

ewjoachim commented:
Ping @reinout ?

reinout commented:
This one seems simpler, so I prefer it.
But: I haven't got experience with version-in-setup.cfg . Does that have to do with the new non-setup.py initiatives?

ewjoachim commented:
Well, I'm not sure what the new non-setup.py initiatives are, but yes, the idea is mainly to put things that should be static configuration in static config files instead of code files.

For an example of how that goes, see Raincoat's [setup.py](https://github.com/novafloss/raincoat/blob/master/setup.py) and [setup.cfg](https://github.com/novafloss/raincoat/blob/master/setup.cfg). All of this is being made possible by setuptool's new (last december) [ability]((https://setuptools.readthedocs.io/en/latest/setuptools.html#configuring-setup-using-setup-cfg-files)) to derive all parameters (sadly, except entrypoint) from setup.cfg.

I'll add the docs from the other PR in this PR.

ewjoachim commented:
Ok, I've updated and rebased the code and docs. I feel like it's missing tests (well, I know it's missing tests, because I tried it on my project and it didn't work at first). I'd be happy to write it, but the existing tests are in a format that, while readable, is quite new to me, and I'm not sure where and how it would be the most appropriate for me to write those.

ewjoachim commented:
Ok, it's updated with tests for the new feature. Ready for review !

reinout commented on doc/source/versions.rst:
```diff
@@ -21,6 +21,12 @@ version number can come from three different locations:
         version='1.0',
         name='...
 
+- The ``setup.cfg`` file. zest.releaser will look for something like:
```
Double colon instead of single one

reinout commented on zest/releaser/tests/vcs.txt:
```diff
@@ -245,3 +245,18 @@ Setup.py with an uppercase VERSION:
+    >>> os.chdir(setupcfgproject)
+    >>> lines = [
+    ...     "[metadata]",
+    ...     "  version = 1.0"]
```
Remove the spaces in front of "version", this breaks the configparser on python 2.7 and on pypy. (That is where the test failure on this PR comes from).

reinout commented on CHANGES.rst:
```diff
@@ -4,7 +4,7 @@ Changelog for zest.releaser
 -------------------
 
-- Nothing changed yet.
+- Write version in ``setup.cfg``. [ewjoachim]
```
Oh, and change this in "Also support version in ``setup.cfg``. "

reinout commented:
Looks good. I've got a few small comments (including a fix for the test failure).

ewjoachim commented:
All fixed. As it was simple fixes, I've taken the step of squashing fixups directly.

Thank you again !

ewjoachim commented on CHANGES.rst:
```diff
@@ -5,6 +5,8 @@ Changelog for zest.releaser
 
 - Nothing changed yet.
+- Also support version in setup.cfg. [ewjoachim]
+- Write version in ``setup.cfg``. [ewjoachim]
```
hm, re-doing this.

ewjoachim replied:
done

reinout commented:
Thanks! I just released 6.12.4, can you check if it all works for you?

ewjoachim commented:
Works great ! On both usual projects and setup.cfg projects. Thank you !
